src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,319 @@
+error details to be included by the server. These are expressed in protocol buffers
+messages, and a
+[set of standard error message types](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+is defined to cover most needs. The protobuf binary encoding of this extra error
+information is provided as trailing metadata in the response.
+
+For more information on the richer error model see the
+[gRPC documentation on error handling](https://grpc.io/docs/guides/error/),
+and the [Google APIs overview of the error model](https://cloud.google.com/apis/design/errors#error_model).
+
+## .NET implementation of the richer error model
+
+The error model is define by the protocol buffers files [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto),
+and the `Google.Api.CommonProtos` NuGet package that provides the generated .NET classes
+from these proto files.
+
+The error is encapsulated by an instance of `Google.Rpc.Status` and
+returned in the trailing response metadata with well-known key `grpc-status-details-bin`.
+Setting and reading this metadata is handled
+for you when using the methods provided in this package.
+
+## Server Side
+
+The server side uses C#'s Object and Collection initializer syntax.
+
+There are two ways that the server can return additional error information:
+- by throwing an `RpcException` that contains the details of the error.
+- by setting the `Status` and adding to the `ResponseTrailers` in the `ServerCallContext`
+before returning from the call. See the examples below.
+
+There are examples of both methods below.  In a .NET client the result will always be
+an exception received by the client no matter which of the above methods the server
+implements.
+
+To add messages to the `Details` repeated field in `Google.Rpc.Status`, wrap each one in `Any.Pack()` - see example below.
+
+The `Google.Rpc.Status` extension method `ToRpcException` creates the appropriate `RpcException` from the status.
+
+The `Grpc.Core.Metadata` extension method `SetRpcStatus` adds a binary representation of the status to the metadata.
+
+
+__Example__ - throwing a `RpcException`:
+```C#
+throw new Google.Rpc.Status
+{
+    Code = Google.Rpc.Code.NotFound;
+    Message = "Simple error message",
+    Details =
+    {
+        Any.Pack(new ErrorInfo
+        {
+            Domain = "error example",
+            Reason = "some reason"
+        }),
+        Any.Pack(new RequestInfo
+        {
+            RequestId = "EchoRequest",
+            ServingData = "Param: " + request.Action.ToString()
+        }),
+        }
+}.ToRpcException();
+```
+__Example__ - setting the status and response trailers instead of throwing an exception:
+```C#
+context.Status = new Grpc.Core.Status(StatusCode.Internal, "Some detail");
+context.ResponseTrailers.SetRpcStatus(new Google.Rpc.Status
```

Comment:
Is there a scenario where a method successfully completes and you would want to set a status in the response trailers?